articles/reproducibility/index.md

```diff
@@ -0,0 +1,115 @@
+
+A second possible solution is the use of timestamps. This method would prove to be more effective than the abovementioned one. Here’s how. The data sources could be designed with timestamps of great accuracy. For example, a database could capture the timestamps. The timestamps could be used to pinpoint training data. It is worth noting that this only works when using databases with the ability to capture said timestamps. If dealing with databases that lack this ability, it may take a lot of effort to redesign them to do so. Plus, this method does not fix the challenge of updating and replacing data in databases.
+
+#### Creation of Features
+
+A parameter that is extracted from data will remain the same in multiple environments only if the data is the same in those environments. Consider a scenario where data has missing values. A common practice to replace a missing value is to have the mean of the variable in focus. However, if training data is different between two environments, this impacts the value of the mean as well.
+
+If [feature extraction](https://deepai.org/machine-learning-glossary-and-terms/feature-extraction) requires complex equations, it makes reproducibility that much harder to achieve. Let’s have a scenario where a feature is the sum total of another feature over time. It is impossible to recalculate such in the absence of all the previous data present in each environment. Furthermore, if features are reliant on random samples, the seed values and random samples have to be spawned in the same order. The seed value has to be the same for each of the environments.
+
+To aid in reproducibility, features that are implemented should not change after being created. They need to immutable. A new feature that is dependent on an already existing one should be represented separately. For example, it should have its own new column. Such measures help in the reduction of dependencies which in turn assists in achieving uniformity across environments. This in turn reduces the potential errors when duplicating a machine learning pipeline in contrasting environments.
+
+#### Model Building
+
+Algorithms themselves pose a challenge to reproducibility. As we mentioned before, many models have an element of randomness in their machine learning process. This randomness has models using the same training data giving different outcomes. For example, [cross validation](/engineering-education/evaluating-ml-model-performance/) may be reliant on random partitioning of data for folds.
+
+The potential solution to this is a simple one; keenness to order. Engineers should be keen on the order in which they pass features, hyperparameters as well as when to introduce [seed values](https://en.wikipedia.org/wiki/Initial_condition).
```

Comment:
Explain seeding as a way to ensure reproducibility. If required, insert a code snippet of numpy using this concept. np,random(seed=#).